### PR TITLE
Pyramid request datadog metrics

### PR DESCRIPTION
**Pull request by paolacastro**

### CODE REVIEW

File: pyramid_datadog/__init__.py [hunk 1]
```diff
@@ -1,20 +1,139 @@
...
+        tags=tags,
+    )
+
+    status_code = str(new_response_event.response.status_code)
+    datadog.increment(
+        'pyramid.request.count',
+        tags=tags + [
+            "status_code:%s" % status_code,
+            "status_type:%sxx" % status_code[0]
+        ],
```
Comment (aconrad) [inline pyramid_datadog/__init__.py]: It seems that `datadog.timing()` already provides the count, I think we don't need this line, we can just use `pyramid.request.duration.total.count` in the UI. Don't forget to move the status code tags above.

File: pyramid_datadog/__init__.py [hunk 2]
```diff
@@ -1,20 +1,139 @@
...
 def on_app_created(app_created_event):
-    datadog = app_created_event.app.registry.datadog
+    registry = app_created_event.app.registry
+    datadog = registry.datadog
     datadog.event(
-        "Pyramid app started",
-        "The Pyramid application has started",
-        tags=["pyramid_datadog"]
+        'Pyramid application %s started' % registry.datadog_app_name,
+        'Pyramid application %s started' % registry.datadog_app_name,
```
Comment (sudarkoff) [inline pyramid_datadog/__init__.py]: Seems redundant. :)

Comment (aconrad) [inline pyramid_datadog/__init__.py]: Yeah, this was meant to be a placeholder until we find what best to put in here.

**Comment by sudarkoff** [comment]
It seems like this code would be sending A LOT of data and might potentially slow down the app significantly. Have you done any benchmarking already?

**Comment by aconrad** [comment]
We are testing on a barebones Pyramid app with a curl hitting the server in a while loop and it looks/feels fine (we don't have anything very scientific though). Although we're tracking down a memory leak right now.

Do you think it would slow it down significantly? I think the library send the metrics to the DD agent on localhost using UDP, so it's fire-and-forget.

**Comment by sudarkoff** [comment]
The library is talking to the agent, yes, but IIRC there's a setting about how much data to aggregate before sending to the server and it's pretty chatty by default. Take a look at your test machine and see how much CPU and traffic your test consumes.

**Comment by aconrad** [comment]
Yes, we'll take a look. But then that wouldn't slow down the app per se, right? (unless you mean it slows it down because the amount of data sent to datadog eats up so much bandwidth that the network has a hard time getting incoming requests).

From what I understand, the DD agent collects metrics locally and aggregates them on a 10s resolution and sends the aggregated data to the server a few times a minute.

This document talks about the DD agent overhead:

> Network: 10-50 KB of bandwidth per minute

source: https://help.datadoghq.com/hc/en-us/articles/203034929-What-is-the-Datadog-Agent-What-resources-does-it-consume-

We're definitely going to benchmark, but I just want to make sure I understand your concern so we are benchmarking the right thing.

**Comment by aconrad** [comment]
@paolacastro you might want to try something like this (untested):

```yaml
language: python

sudo: false

matrix:
  include:
    - python: 2.7
      env: TOXENV=py27
    - python: 3.5
      env: TOXENV=py35

install:
  - travis_retry pip install tox

script:
  - tox
```

**Comment by aconrad** [comment]
I made the change myself and it worked! We can now wire up couverture.io

**Comment by aconrad** [comment]
Ok, I finally wired-up couverture.io for coverage reports. It was tedious because Travis makes it non-obvious on how to get the real commit SHA when the build is running within their pull request pipeline.

File: pyramid_datadog/__init__.py (hunk 2 above)
Comment (aconrad) [inline pyramid_datadog/__init__.py]: We'll likely try to extract the version number in here, but I think we'll do this in another PR.

File: tests/test_includeme.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+        mock.call(on_new_response, NewResponse),
+    ])
+
+
+def test_configure_metrics():
+    config = mock.Mock()
+    datadog_metrics = mock.Mock()
+    configure_metrics(config, 'app_name', datadog_metrics)
+
+    config.registry.datadog == datadog_metrics
```
Comment (aconrad) [inline tests/test_includeme.py]: you need an `assert` keyword here

File: tests/test_includeme.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+            tags=['app:app_name', 'route:test_route', 'status_code:200', 'status_type:2xx']
+        ),
+    ])
+
+
+@patch('time.time')
+def test_time_ms(mock_time):
+    mock_time.return_value = 1
+    return_value = time_ms()
+    assert return_value == 1000
```
Comment (aconrad) [inline tests/test_includeme.py]: Nice, this ensures milliseconds are returned!

File: pyramid_datadog/__init__.py [hunk 5]
```diff
@@ -1,20 +1,137 @@
...
+
+
+def configure_metrics(config, app_name, datadog_metrics):
+    '''
+    * app_name: name of your application that will be in a tag
+    * datadog_metrics: datadog metrics object initialized by user
+    '''
     config.registry.datadog = datadog_metrics
+    config.registry.datadog_app_name = app_name
+    config.registry.datadog_app_tag = 'app:%s' % app_name
```
Comment (aconrad) [inline pyramid_datadog/__init__.py]: For some reason this `datadog_app_tag` rubs me the wrong way. I think we might be a little too opinionated by automatically adding a tag `"app:myapp"` here. What if users have conventions about how they do tagging and ours wouldn't follow theirs?

The more I think about it the more I think we should let users provide a list of tags they want add to every metric emitted. For example:

```python
config.configure_metrics(datadog, tags=["application:foobar"])
```

... this way every metric generated would also be tagged with the user's tags.

Comment (sudarkoff) [inline pyramid_datadog/__init__.py]: For consistency, we need to have reserved tags that apply across the board, otherwise we won't be able to aggregate across all services. So, `application` (spelled out fully) is a really good example of such standard tag. IF it's different from the role, that is. Because `role:<service_name>` will already be added by the infra.

Comment (aconrad) [inline pyramid_datadog/__init__.py]: Yup, but this is an open source lib so we're not going to add our opinionated ideas in here. We have our own lib to do what you are describing ;)

**Comment by aconrad** [comment]
Actually, let's remove the tags that a user can set upon initialization for now... We (or the open source community) can implement them if we see a need for them (which we currently don't). I want to limit the exposure of public APIs as much as possible until we see how things play out and not based on assumptions.

File: pyramid_datadog/__init__.py (hunk 5 above)
Comment (sudarkoff) [inline pyramid_datadog/__init__.py]: Ah, good point.

**Comment by sudarkoff** [comment]
I don't understand the motivation for removing the tags. Why not just make them optional? Metrics without tags would be very difficult to use.

**Comment by aconrad** [comment]
Because the library we are going to use as our Datadog agent (not open source) already has this feature and I'd like to oil the gears a bit before exposing public APIs that I'm unsure how they are going to be used (a public API is like a contract, you screw it up you can't get rid of it without introducing backwards incompatible changes). If the need arises, it would literally be a 10-liner PR, it's a quick improvement.
